File: train_mnist.py
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader
from torchvision import datasets, transforms

# --- Hyperparameters ---
LEARNING_RATE = 1e-3
BATCH_SIZE = 64
EPOCHS = 10


# --- Model ---
class MNISTNet(nn.Module):
    def __init__(self):
        super().__init__()
        self.features = nn.Sequential(
            nn.Conv2d(1, 32, kernel_size=3, padding=1),  # 28x28
            nn.ReLU(),
            nn.MaxPool2d(2),                              # 14x14
            nn.Conv2d(32, 64, kernel_size=3, padding=1), # 14x14
            nn.ReLU(),
            nn.MaxPool2d(2),                              # 7x7
        )
        self.classifier = nn.Sequential(
            nn.Flatten(),
            nn.Linear(64 * 7 * 7, 128),
            nn.ReLU(),
            nn.Dropout(0.25),
            nn.Linear(128, 10),
        )

    def forward(self, x):
        return self.classifier(self.features(x))


def train_epoch(model, loader, optimizer, criterion, device):
    model.train()
    total_loss, correct = 0.0, 0
    for X_batch, y_batch in loader:
        X_batch, y_batch = X_batch.to(device), y_batch.to(device)
        optimizer.zero_grad()
        preds = model(X_batch)
        loss = criterion(preds, y_batch)
        loss.backward()
        optimizer.step()
        total_loss += loss.item() * len(X_batch)
        correct += (preds.argmax(1) == y_batch).sum().item()
    n = len(loader.dataset)
    return total_loss / n, correct / n


def eval_epoch(model, loader, criterion, device):
    model.eval()
    total_loss, correct = 0.0, 0
    with torch.no_grad():
        for X_batch, y_batch in loader:
            X_batch, y_batch = X_batch.to(device), y_batch.to(device)
            preds = model(X_batch)
            total_loss += criterion(preds, y_batch).item() * len(X_batch)
            correct += (preds.argmax(1) == y_batch).sum().item()
    n = len(loader.dataset)
    return total_loss / n, correct / n


if __name__ == "__main__":
    transform = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,)),  # MNIST mean/std
    ])

    train_set = datasets.MNIST("data", train=True, download=True, transform=transform)
    val_set = datasets.MNIST("data", train=False, download=True, transform=transform)

    train_loader = DataLoader(train_set, batch_size=BATCH_SIZE, shuffle=True, num_workers=2)
    val_loader = DataLoader(val_set, batch_size=BATCH_SIZE, num_workers=2)

    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    print(f"Using device: {device}")

    model = MNISTNet().to(device)
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.Adam(model.parameters(), lr=LEARNING_RATE)

    for epoch in range(1, EPOCHS + 1):
        train_loss, train_acc = train_epoch(model, train_loader, optimizer, criterion, device)
        val_loss, val_acc = eval_epoch(model, val_loader, criterion, device)
        print(
            f"Epoch {epoch:>2}/{EPOCHS}  "
            f"train_loss={train_loss:.4f}  train_acc={train_acc:.4f}  "
            f"val_loss={val_loss:.4f}  val_acc={val_acc:.4f}"
        )

    print("\nTraining complete.")
    torch.save(model.state_dict(), "mnist_model.pth")
    print("Model saved to mnist_model.pth")

File: train_mnist.cpp
#include <torch/torch.h>
#include <iostream>
#include <iomanip>

// --- Hyperparameters ---
const int BATCH_SIZE = 64;
const int EPOCHS = 10;
const float LEARNING_RATE = 1e-3;

// --- Model ---
struct MNISTNet : torch::nn::Module {
    torch::nn::Sequential features, classifier;

    MNISTNet()
        : features(
              torch::nn::Conv2d(torch::nn::Conv2dOptions(1, 32, 3).padding(1)),
              torch::nn::ReLU(),
              torch::nn::MaxPool2d(2),
              torch::nn::Conv2d(torch::nn::Conv2dOptions(32, 64, 3).padding(1)),
              torch::nn::ReLU(),
              torch::nn::MaxPool2d(2)),
          classifier(
              torch::nn::Flatten(),
              torch::nn::Linear(64 * 7 * 7, 128),
              torch::nn::ReLU(),
              torch::nn::Dropout(0.25),
              torch::nn::Linear(128, 10))
    {
        register_module("features", features);
        register_module("classifier", classifier);
    }

    torch::Tensor forward(torch::Tensor x) {
        return classifier->forward(features->forward(x));
    }
};

// --- Train one epoch ---
template <typename Loader>
std::pair<double, double> train_epoch(
    MNISTNet& model,
    Loader& loader,
    torch::optim::Adam& optimizer,
    torch::Device device)
{
    model.train();
    double total_loss = 0.0;
    int correct = 0, total = 0;

    for (auto& batch : loader) {
        auto data = batch.data.to(device);
        auto targets = batch.target.to(device);

        optimizer.zero_grad();
        auto output = model.forward(data);
        auto loss = torch::nn::functional::cross_entropy(output, targets);
        loss.backward();
        optimizer.step();

        total_loss += loss.item().toDouble() * data.size(0);
        correct += output.argmax(1).eq(targets).sum().item().toInt();
        total += data.size(0);
    }
    return {total_loss / total, static_cast<double>(correct) / total};
}

// --- Eval one epoch ---
template <typename Loader>
std::pair<double, double> eval_epoch(
    MNISTNet& model,
    Loader& loader,
    torch::Device device)
{
    model.eval();
    torch::NoGradGuard no_grad;
    double total_loss = 0.0;
    int correct = 0, total = 0;

    for (auto& batch : loader) {
        auto data = batch.data.to(device);
        auto targets = batch.target.to(device);

        auto output = model.forward(data);
        auto loss = torch::nn::functional::cross_entropy(output, targets);

        total_loss += loss.item().toDouble() * data.size(0);
        correct += output.argmax(1).eq(targets).sum().item().toInt();
        total += data.size(0);
    }
    return {total_loss / total, static_cast<double>(correct) / total};
}

int main() {
    torch::Device device = torch::cuda::is_available()
        ? torch::Device(torch::kCUDA)
        : torch::Device(torch::kCPU);
    std::cout << "Using device: " << device << "\n";

    // --- Data ---
    auto train_dataset = torch::data::datasets::MNIST("data")
        .map(torch::data::transforms::Normalize<>(0.1307, 0.3081))
        .map(torch::data::transforms::Stack<>());

    auto val_dataset = torch::data::datasets::MNIST("data", torch::data::datasets::MNIST::Mode::kTest)
        .map(torch::data::transforms::Normalize<>(0.1307, 0.3081))
        .map(torch::data::transforms::Stack<>());

    auto train_loader = torch::data::make_data_loader<torch::data::samplers::RandomSampler>(
        std::move(train_dataset), torch::data::DataLoaderOptions().batch_size(BATCH_SIZE));

    auto val_loader = torch::data::make_data_loader<torch::data::samplers::SequentialSampler>(
        std::move(val_dataset), torch::data::DataLoaderOptions().batch_size(BATCH_SIZE));

    // --- Model + optimizer ---
    MNISTNet model;
    model.to(device);
    torch::optim::Adam optimizer(model.parameters(), torch::optim::AdamOptions(LEARNING_RATE));

    // --- Training loop ---
    for (int epoch = 1; epoch <= EPOCHS; ++epoch) {
        auto [train_loss, train_acc] = train_epoch(model, *train_loader, optimizer, device);
        auto [val_loss, val_acc] = eval_epoch(model, *val_loader, device);

        std::cout << std::fixed << std::setprecision(4)
                  << "Epoch " << std::setw(2) << epoch << "/" << EPOCHS
                  << "  train_loss=" << train_loss
                  << "  train_acc=" << train_acc
                  << "  val_loss=" << val_loss
                  << "  val_acc=" << val_acc << "\n";
    }

    torch::save(std::make_shared<MNISTNet>(model), "mnist_model_cpp.pt");
    std::cout << "\nTraining complete. Model saved to mnist_model_cpp.pt\n";
    return 0;
}
